### DISCUSSION TITLE
Custom iso8601 config with serde

### DISCUSSION THREAD

**Discussion by TristanDebrunner** [time-rs/time#555] [Question] [answered]
In #548, there is mention of the ability to customize the iso8601 parsing with `#[serde(with)]` in [this comment](https://github.com/time-rs/time/issues/548#issuecomment-1419760669):

> With regard to using a different configuration, the next release will be helpful to you. You'll be able to use something like the following:
> 
> ```rust
> use time::format_description::well_known::{Iso8601, iso8601};
> const FORMAT: _ = Iso8601<{ iso8601::Config::DEFAULT.set_year_is_six_digits(false).encode() }>;
> time::serde::format_description!(my_format, OffsetDateTime, FORMAT);
> ```

There appear to have been 3 releases since that comment, but I cannot find any documentation about this kind of customization, and trying variations on the above code always causes errors. Is this capability still a work in progress?

**Comment by jhpratt** [comment, marked as answer]
Ultimately I need to improve documentation. I'll check when I get on my computer, but this behavior _should_ be possible on the latest release.

**Reply by TristanDebrunner** [reply]
The main problem seems to be that the type of `FORMAT` is `Iso8601<N>` where `N` is some large number, but `serde::format_description!()` expects either a string literal or `&[FormatItem<'_>]` (or a path to such). I haven't been able to find a solution or workaround to the type mismatch.

If there is a solution to this, I'd be happy to PR some doc changes with the details.

**Reply by jhpratt** [reply]
Looking at the expansion, it's because I neglected to consider that the format could be something other than `&[FormatItem<'_>]` when making the change. I believe there is an expansion that will make this work, even if it does have to work around a limitation of the compiler.

For reference, this is the exact code I am using:

```rust
use time::format_description::well_known::{iso8601, Iso8601};
const FORMAT: Iso8601<
    {
        iso8601::Config::DEFAULT
            .set_year_is_six_digits(false)
            .encode()
    },
> = Iso8601::<
    {
        iso8601::Config::DEFAULT
            .set_year_is_six_digits(false)
            .encode()
    },
>;
time::serde::format_description!(my_format, OffsetDateTime, FORMAT);
```

The reason you're seeing a large, seemingly arbitrary number is because that's what the encoded configuration looks like. It's a complete hack, but it works.

**Reply by jhpratt** [reply]
I've pushed up a change that ensures it works regardless of type, so long as it implements `Formattable` and/or `Parsable` (depending on feature flags).

**Reply by TristanDebrunner** [reply]
Thanks for making that fix. It works when I pull from main now. I'll look forward to the next release.

For reference, this code also works and reduces repetition:
```rust
use time::format_description::well_known::{iso8601::{self, EncodedConfig}, Iso8601};
const CONFIG: EncodedConfig = iso8601::Config::DEFAULT
    .set_year_is_six_digits(false)
    .encode();
const FORMAT: Iso8601<CONFIG> = Iso8601::<CONFIG>;
time::serde::format_description!(my_format, OffsetDateTime, FORMAT);
```
